File: solutions/clone-graph/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        oldNewMap = {}
        def cloneNode(node):
            if node in oldNewMap:
                return oldNewMap[node]
            
            copy = Node(node.val)
            oldNewMap[node] = copy
            for neighbor in node.neighbors:
                copy.neighbors.append(cloneNode(neighbor))
            return copy
        return cloneNode(node) if node else None

File: solutions/clone-graph/clone-graph.go
/**
 * Definition for a Node.
 * type Node struct {
 *     Val int
 *     Neighbors []*Node
 * }
 */

func cloneGraph(node *Node) *Node {
    oldToNew := map[*Node]*Node{} 
    if node == nil {
        return nil
    }
    return cloneNode(node, &oldToNew)
}

func cloneNode(n *Node, m *map[*Node]*Node ) *Node {
    if val, ok := (*m)[n]; ok {
        return val
    }
    copy := &Node{
        Val: n.Val,
        Neighbors: []*Node{},
    }
    (*m)[n] = copy
    for _, neighbor := range (*n).Neighbors {
        copy.Neighbors = append(copy.Neighbors, cloneNode(neighbor, m))
    }
    return copy
}